rfcs/2021-12-01-9460-health-endpoint-improvements.md

```diff
@@ -0,0 +1,117 @@
+# RFC 9460 - 2021-12-01 - Health endpoint improvements
+
+Our existing `/health` endpoint is limited and additional information should be
+exposed for both operators and load balancers/service discovery.
+
+## Context
+
+- [Additional response for api health endpoint](https://github.com/vectordotdev/vector/issues/9160)
+- [Formalize component health to include statuses and runtime health](https://github.com/vectordotdev/vector/issues/10555)
+- [Add specification for the `/health` endpoint](https://github.com/vectordotdev/vector/issues/10556)
+- [Add component health status the `/heath` endpoint](https://github.com/vectordotdev/vector/issues/9469)
+
+## Scope
+
+### In scope
+
+- Update health endpoint to improve Vector's reliability in production deployments
+- Update default configurations to improve Vector's out-of-the-box reliability
+- Improve visibility of high level status at runtime
+
+### Out of scope
+
+- Immediate integrations to expose component level healh
+
+## Pain
+
+- Today's health endpoint causes 502's when running behind a loadbalancer
+- The healthcheck isn't integrated with any components and only represents if Vector
+itself is running
+- Operators don't have visibility into the state of specific components, ex. is
+backpressure being applied to upstream components
+
+## Proposal
+
+### User Experience
+
+The existing health endpoint on Vector's API is updated to return 503's
+when Vector is shutting down. This additional response will allow load balancers
+and service discovery to better integrate with Vector, sources already begin
+rejecting requests received during the shutdown process and updating the health
+check will allow for removing the instance from available backends and avoid
+routing traffic to an instance that will end up rejecting the request regardless.
+
+The API will not start until the topology is build and validated as functional,
+this isn't precisely when Vector's configured sources _can_ actually process
+events but it's a reasonable first step to ensure graceful startups.
```

Comment:
This makes me think that propagating source's health is the correct solution since both are dependent on sources and sources already implement startup/shutdown logic.

Reply:
Yes, that is part of the future work and specification definition.